### PR TITLE
IS-1089: Write the log of precommit data to the seperated file

### PR DESCRIPTION
**Pull request by soobokjin**
## Summary
- format: JSON
- file path & file name: {LOG PATH}/precommit/{block height}-precommit-data-v{version num}.json
- Add tx_index field to TransactionBatchValue for debugging
  -  There is some issue about retrieve index from the context in case of deploy (In deploy, context.tx is set to None temporally)
- Check assert when set items (i.e. __setitem__) on TransactionBatch (must be TransactionBatchValue)
  - Only TransactionBatchValue type can be set on TransactionBatch

## Information
- I have left the comments on the major change point

## Format 
```
{
  "iconservice": "1.6.1",
  "revision": 5,
  "block": {
    "height": 2,
    "hash": "0xbf1ba1fcc5894f900a2424c7c23f5b4df52ddfa512aa6611554cffa63af3c885",
    "timestamp": 1591318988717094,
    "prevHash": "0x7833cedff6814fef8dc5f818d6307ef537e6d79087620629b73de8bf91caa283",
    "cumulativeFee": 1318800000000000
  },
  "isStateRootHash": "0x44ee9cd260abf6ff2be79ef10d8bfc8ae7a4da32e5fe4bb850873f7060dc2453",
  "rcStateRootHash": "0x44ee9cd260abf6ff2be79ef10d8bfc8ae7a4da32e5fe4bb850873f7060dc2453",
  "stateRootHash": "0x44ee9cd260abf6ff2be79ef10d8bfc8ae7a4da32e5fe4bb850873f7060dc2453",
  "prevBlockGenerator": null,
  "blockBatch": [
    {
      "key": "0x0100000000000000000000000000000000000000017c027c7265766973696f6e5f636f6465",
      "value": "0x05",
      "includeStateRootHash": true,
      "txIndexes": [
        0
      ]
    },
    ..
    {
      "key": "0x696973737272",
      "value": "0x9200cd04b0",
      "includeStateRootHash": true,
      "txIndexes": [
        0
      ]
    }
  ],
  "rcBlockBatch": [
    {
      "key": "0x4844",
      "value": "0x920002"
    }
  ]
}
```

### CODE REVIEW

File: iconservice/database/batch.py [hunk 1]
```diff
@@ -88,6 +142,8 @@ def __getitem__(self, item):
         return None
 
     def __setitem__(self, key, value):
+        assert isinstance(value, TransactionBatchValue)
```
Comment (soobokjin) [inline iconservice/database/batch.py:145]: I have added the assert for reducing confusion

File: iconservice/database/batch.py [hunk 2]
```diff
@@ -159,8 +216,25 @@ def __init__(self, block: Optional['Block'] = None):
...
+        :return:
+        """
+        new_list: list = []
+        for key, val in self.items():
+            _dict = {"key": key}
+            _dict.update(val.to_dict(to_camel_case))
+            new_list.append(_dict)
+        return new_list
+
     def update(self, tx_batch: 'TransactionBatch', **kwargs):
```
Comment (soobokjin) [inline iconservice/database/batch.py:231]: If the same key has been put on block_batch, pop previous value for extract the tx index

File: iconservice/database/db.py [hunk 3]
```diff
@@ -273,7 +273,8 @@ def _put(self,
             self.key_value_db.put(key, value)
         else:
             self._check_tx_batch_value(context, key, include_state_root_hash)
-            context.tx_batch[key] = TransactionBatchValue(value, include_state_root_hash)
+            tx_index: int = context.tx.index if context.tx is not None else -1
```
Comment (soobokjin) [inline iconservice/database/db.py:276]: Get index info from context.tx. if None, set as -1

File: iconservice/database/batch.py [hunk 4]
```diff
@@ -15,16 +15,69 @@
...
 
 from ..base.block import Block
 from ..base.exception import DatabaseException, AccessDeniedException
 from ..icx import IcxStorage
-from ..utils import sha3_256
+from ..utils import sha3_256, to_camel_case
 
-TransactionBatchValue = namedtuple('TransactionBatchValue', ['value', 'include_state_root_hash'])
+
+class TransactionBatchValue:
```
Comment (eunsoo-icon) [inline iconservice/database/batch.py:28]: Confused class name .... BlockBatchValue seems to be better.

File: iconservice/database/batch.py [hunk 5]
```diff
@@ -15,16 +15,92 @@
...
+        """
+        # Fixme: Correct tx index should be set on deploying score (not -1)
+        return self._tx_index
+
+    @tx_index.setter
+    def tx_index(self, _):
+        raise AccessDeniedException(f"Cannot set the tx index")
+
+    def __repr__(self):
+        return 'TransactionBatchValue(%s, %d, %s)' % (self.value.hex(), self.include_state_root_hash, self.tx_index)
```
[the remark this answers is not in the document]
Comment (soobokjin) [reply iconservice/database/batch.py]: I have changed to using f'' instead.

**Pull request merged**